### ISSUE TITLE
Taxonomy Support (Tags, Categories)

### ISSUE THREAD

**Issue by jauyong** [GoogleForCreators/web-stories-wp#489]
## Feature Description

- Adding new tab to the `Document` panel of the editor that allows users to set tags and categories of their story. 
- This will mimic the ability to add, edit, delete tags and categories to a story that users have for traditional posts in  WordPress. 

UI draft from Aug 28, 2021 (where the part on the leftt is what Gutenberg (WordPress) does now and the bit on the right is to implement in the editor): 
![image](https://user-images.githubusercontent.com/10720454/131265367-f15bc533-38ff-46c9-bdd7-52f42fa0f31b.png)

_______ 
(March 4, 2020 - original description)
This was taken out of #405 

**[Feature Brief](https://docs.google.com/document/d/1GJZljicdXcP7oRdywsF3EGUaCVpCHsDrK2gUMubnrbk/edit?pli=1#heading=h.3oygdma1cd8e)**

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria
1. Authors are able to add a new category
10. Authors are able to add a new tag
8. Authors are able to edit the categories

**Comment by miina** [comment]
Couldn't find design currently in Figma for this, assuming it should be in the Document panel.
Probably UX Needed?

@o-fernandez -- adding UX Needed label, LMK if the design actually exists and I missed it, so we can add it to the description as well.

cc @samitron7

**Comment by o-fernandez** [comment]
@pbakaus @swissspidy Want your thoughts here. IIRC, our main way of getting the story into the site's various feeds and distribution will be by having users embed their story into a post they publish. Is that still the plan? 

If so, then the priority of this could possibly drop, since the article/post where the story is embedded could be where the right categories/tags are added. Let me know what you think.

**Comment by swissspidy** [comment]
> our main way of getting the story into the site's various feeds and distribution will be by having users embed their story into a post they publish. Is that still the plan?

Yes. I think this would be more about helping publishers organize their content in the admin rather than having to do anything with distribution.

**Comment by o-fernandez** [comment]
Great, thanks for confirming @swissspidy. I think that makes sense, and definitely important to do, but lowers the urgency in my mind. Let me discuss prioritization with Sam in our upcoming 1-1.

**Comment by swissspidy** [comment]
Sounds good.

To give some more context:

Right now stories already support adding tags & categories via the "All Stories" view (aka classic WordPress list table):

<img width="1349" alt="Screenshot 2020-05-18 at 16 01 43" src="https://user-images.githubusercontent.com/841956/82221923-eb113c80-9920-11ea-83ae-d604e3d4f0b5.png">

They're also already shown in the Dashboard:

<img width="1510" alt="Screenshot 2020-05-18 at 15 48 21" src="https://user-images.githubusercontent.com/841956/82222009-02502a00-9921-11ea-9664-106ce1c6645d.png">

Maybe that's a bit confusing for users?

So this issue would only be about adding them to the Document panel in the editor.

Another interesting aspect: right now the tags and categories are shared with regular posts. So if you have the tag "Awesome", it can be added to both a post and a story. However, we could also make them unique to stories, and not shared with posts.

**Comment by o-fernandez** [comment]
Open questions: 
* do users need both categories and tags to organize their stories? 
* do they want to use the same categories/tags as they use for the rest of WP, or story-specific tagging? 

If we feel comfortable saying that we have answers for these, then we could build this now. Otherwise, do we have the option to hide these from the dashboard and "all stories" sections until we have user feedback/need to determine how we should implement this? 

@swissspidy @pbakaus @carlos-kelly

**Comment by swissspidy** [comment]
> * do users need both categories and tags to organize their stories?

I don't think so. Even in WordPress itself many users find it confusing to have both options.

IMO 1 option - tags - is enough.

> * do they want to use the same categories/tags as they use for the rest of WP, or story-specific tagging?

Sharing tags can cause issues in WP, as the counts for how often a tag is used can be wrong inn this case. Personally I'd argue for not doing this sharing.
If someone really wants it, they could build a small WordPress plugin to do so.


> Otherwise, do we have the option to hide these from the dashboard and "all stories" sections until we have user feedback/need to determine how we should implement this?

Yes we do. Should be relatively simple to make this change.

My suggestion: let's do that until we have user feedback. It's not a V1 feature anyway.

**Comment by o-fernandez** [comment]
Thanks @swissspidy - that sounds good! Moving this out of the Beta/Stable release. Filed #1884 for now.

**Comment by swissspidy** [comment]
There was a user request for this at #3107

**Comment by swissspidy** [comment]
This one is quite likely needed for the upcoming block library work. Users need a way to categorize stories so they can display stories on their websites for particular topics.

We can start with tags only for now, and have them _not_ shared with posts.

The feature should be built in an agnostic way independent from specific implementations.

In other words, if we ever add a "Country", "Color" or "Whatever" taxonomy, there should be no changes needed to the editor or dashboard.

@o-fernandez WDYT?

So we need:

* a short spec for this
* some UX work for how this looks like in editor / dashboard

Not 100% blocked by UX. For the time being users could just use the "All Stories" view.

**Comment by o-fernandez** [comment]
@swissspidy agree that this will be important for the blocks. Curious, though, why would we make this independent from posts? My thinking is - categories and tags are probably something publishers set up to match the different page experiences they want to create, so asking them to set a new taxonomy for these posts may not be great. But, wondering if there are WP reasons for which you recommend not doing this? 

@amedina and I had discussed this ticket a few weeks ago when talking about embeds and agreed it would be important to do this as part of the enhanced blocks. I think @amedina had some thoughts on how this could work, so may be good to sync on it. 

My short spec would be (I can put it in a brief)

1. Allow user to add tags and categories in the "document" tab in the editor, mimicking how it works for posts/pages except in our editor UI. 
2. I would have recommended to use the same category and tags taxonomy as the rest of the site, but open to change this. 
3. User should be able to modify this categories and tags in the "All Posts/classic view" 
4. We can file a separate ticket to surface tags and categories in the dashboard and let people modify them from there, to not force features into the classic "all posts" view. But, non-blocking to ship the feature.

If this generally sounds good, we can file tickets for 1 (workspace change) and 4 (dashboard change), but LMKWYT.

**Comment by swissspidy** [comment]
I suppose we'll need to refine what the goal of this is.

Is it purely for managing content in the backend, or also for showing stories of these categories in the archives? For example, WP by default has archives like `example.com/category/my-awesome-category/` that lists posts of this category only. It would not display stories out off the box (unless we'd use something like https://github.com/spacedmonkey/built-in-taxos-cpt)

> I would have recommended to use the same category and tags taxonomy as the rest of the site, but open to change this.

One issue with that is that WordPress will have troubles counting how many content items have a specific category, because by default it only counts posts, not custom post types like stories.

Another issue is the archives situation above.

**Comment by dreamofabear** [comment]
Bumping due to a FR from NCA, which consists of two user flows:
1. [P0] Allow registering/associating an existing taxonomy with the `web-story` post type (on the Settings page?)
2. [P1] Allow setting a story with a registered taxonomy in the editor (in the Document panel?) and displaying existing associated taxonomy values 

I think this is pretty close to the existing proposals above, just generalized.

**Comment by swissspidy** [comment]
> 1. [P0] Allow registering/associating an existing taxonomy with the `web-story` post type (on the Settings page?)

This is not something we should add to the settings page. This belongs to where taxonomies are registered.

Wherever NCA is registering custom taxonomies in their plugin(s), they should just update their `register_taxonomy()` call or use `register_taxonomy_for_object_type()` accordingly. Examples:

```php
register_taxonomy( 'my-cool-taxonomy', [ 'post', 'web-story', 'some-other-post-type' ];

register_taxonomy_for_object_type( 'my-cool-taxonomy', 'web-story' );
```


> 2. [P1] Allow setting a story with a registered taxonomy in the editor (in the Document panel?) and displaying existing associated taxonomy values
> 

This ticket's goal, essentially.

`register_taxonomy()` allows registering two types of taxonomies, hierarchical (e.g. categories with sub-categories) and non-hierarchical ones (e.g. tags).

We we would take the list of all taxonomies registered for the web-story post type and display them in the document panel. The UI component being used depends on whether it's hierarchical or not.

This is how this works in WP elsewhere and what users and developers would expect here as well.

Example UI for hierarchical and non-hierarchical taxonomies in Gutenberg:

![Screenshot 2021-06-24 at 08 51 45](https://user-images.githubusercontent.com/841956/123216042-6a305700-d4c9-11eb-8839-3f5ab09a8e52.png)

**Comment by bmattb** [comment]
requires UX direction on a multi-select dropdown,
